Objects/unicodeobject.c

```diff
@@ -10673,6 +10673,67 @@ PyUnicode_CompareWithASCIIString(PyObject* uni, const char* str)
     }
 }
 
+int
+PyUnicode_EqualToUTF8(PyObject *unicode, const char *str)
+{
+    assert(_PyUnicode_CHECK(unicode));
+    assert(str);
+    if (PyUnicode_IS_ASCII(unicode)) {
+        size_t len = (size_t)PyUnicode_GET_LENGTH(unicode);
+        return strlen(str) == len &&
+            memcmp(PyUnicode_1BYTE_DATA(unicode), str, len) == 0;
+    }
+    if (PyUnicode_UTF8(unicode) != NULL) {
+        size_t len = (size_t)PyUnicode_UTF8_LENGTH(unicode);
+        return strlen(str) == len &&
+            memcmp(PyUnicode_UTF8(unicode), str, len) == 0;
+    }
+
+    Py_UCS4 ch;
+    Py_ssize_t len = PyUnicode_GET_LENGTH(unicode);
+    int kind = PyUnicode_KIND(unicode);
+    const void *data = PyUnicode_DATA(unicode);
+    /* Compare Unicode string and UTF-8 string */
+    for (Py_ssize_t i = 0; i < len; i++) {
+        ch = PyUnicode_READ(kind, data, i);
+        if (ch == 0) {
+            return 0;
+        }
+        else if (ch < 0x80) {
+            if (ch != (unsigned char)*str++) {
+                return 0;
+            }
+        }
+        else if (ch < 0x800) {
+            if ((0xc0 | (ch >> 6)) != (unsigned char)*str++ ||
+                (0x80 | (ch & 0x3f)) != (unsigned char)*str++)
```

Comment:
```
unsigned char byte1 = (0xc0 | (ch >> 6));
unsigned char byte2 = (0x80 | (ch & 0x3f));
if (str[0] != byte1 || str[1] != byte2) return 0;
```

And declare a *str* variable as ``unsigned char*`` once to avoid casting str at each byte comparison.

Reply:
If the first comparison fails, you do not need to calculate the second byte. The code looks more compact and uniform in the way it is written right now. All expressions I copied from the UTF-8 encoder which I wrote 11 years ago, so no need to recheck them. Casting to unsigned char is not a large burden, but if you prefer, I can introduce a new `unsigned char*` variable.